source/Server/IntegratedAuthentication/IntegratedAuthenticationHandler.cs

```diff
@@ -0,0 +1,122 @@
+                    state = JsonConvert.DeserializeObject<LoginState>(stateString.FirstOrDefault());
+                }
+                catch (Exception e)
+                {
+                    log.Warn(e, "Invalid login state object passed to the server when setting up the NTLM challenge. Falling back to the default behaviour.");
+                }
+            }
+
+            // Build the auth cookies to send back with the response
+            var authCookies = tokenIssuer.CreateAuthCookies(principal.IdentificationToken, SessionExpiry.TwentyDays, context.Request.IsHttps, state?.UsingSecureConnection);
+
+            // If the caller has provided a redirect after successful login, we need to check it is a local address - preventing Open Redirection attacks
+            if (!string.IsNullOrWhiteSpace(state?.RedirectAfterLoginTo))
+            {
+                var whitelist = authenticationConfigurationStore.GetTrustedRedirectUrls();
+                var redirectAfterLoginTo = state.RedirectAfterLoginTo.Replace(DirectoryServicesConstants.IntegratedAuthVirtualDirectory, string.Empty);
+                if (Requests.IsLocalUrl(redirectAfterLoginTo, whitelist))
+                {
+                    // This is a safe redirect, let's go!
+                    context.Response.Redirect(redirectAfterLoginTo);
+                    foreach (var cookie in authCookies)
+                    {
+                        context.Response.Cookies.Append(cookie.Name, cookie.Value);
+                    }
+                    return Task.CompletedTask;
+                }
+
+                // Just log that we detected a non-local redirect URL, and fall through to the root of the local web site
+                log.WarnFormat(
+                    "Prevented potential Open Redirection attack on an NTLM challenge, to the non-local url {0}",
+                    redirectAfterLoginTo);
+            }
+
+            // By default, redirect to the root of the local web site
+            context.Response.Redirect(context.Request.PathBase.Value ?? "/");
+            foreach (var cookie in authCookies)
+            {
+                context.Response.Cookies.Append(cookie.Name, cookie.Value);
+            }
+            
+            return Task.CompletedTask;
+        }
+        
+        OctopusAuthenticationResult TryAuthenticateRequest(HttpContext context)
+        {
+            if (context == null) throw new ArgumentNullException(nameof(context));
+            if (context.Request == null) throw new ArgumentNullException($"{nameof(context)}.{nameof(context.Request)}");
+            
+            // If there is no "RequestPrincipal" in the Context.Items it's not our job to authenticate this request
+            var principal = context.User;
+            if (string.IsNullOrWhiteSpace(principal.Identity.Name))
+                return OctopusAuthenticationResult.Anonymous;
+
+            using (var cts = new CancellationTokenSource(TimeSpan.FromMinutes(1)))
```

Comment:
The `HttpContext` has https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.httpcontext.requestaborted?view=aspnetcore-3.1#Microsoft_AspNetCore_Http_HttpContext_RequestAborted, I wonder if we should use `CancellationTokenSource.CreateLinkedTokenSource(context.RequestAborted, new CancellationTokenSource(TimeSpan.FromMinutes(1)).Token)`